File: app/models/store.py
"""Store model and the controlled first-version platform values."""

from datetime import datetime
from enum import Enum
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, Enum as SqlEnum, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base

if TYPE_CHECKING:
    from app.models.product import Product


class Platform(str, Enum):
    TAOBAO = "taobao"
    TMALL = "tmall"
    JD = "jd"
    DOUYIN = "douyin"
    PINDUODUO = "pinduoduo"
    OTHER = "other"


class Store(Base):
    __tablename__ = "stores"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    store_name: Mapped[str] = mapped_column(String(100), nullable=False)
    platform: Mapped[Platform] = mapped_column(
        SqlEnum(
            Platform,
            name="platform_type",
            native_enum=False,
            create_constraint=True,
            validate_strings=True,
            values_callable=lambda enum_class: [item.value for item in enum_class],
        ),
        nullable=False,
        index=True,
    )
    external_store_id: Mapped[str | None] = mapped_column(String(100))
    owner_name: Mapped[str | None] = mapped_column(String(100))
    remark: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False,
    )

    products: Mapped[list["Product"]] = relationship(
        back_populates="store",
        passive_deletes=True,
    )

File: frontend/src/types/store.ts
export type Platform = 'taobao' | 'tmall' | 'jd' | 'douyin' | 'pinduoduo' | 'other'

export interface Store {
  id: number
  store_name: string
  platform: Platform
  external_store_id: string | null
  owner_name: string | null
  remark: string | null
  created_at: string
  updated_at: string
}

export interface StoreCreate {
  store_name: string
  platform: Platform
  external_store_id?: string | null
  owner_name?: string | null
  remark?: string | null
}

export type StoreUpdate = Partial<StoreCreate>
